File: enter.pollinations.ai/scripts/user-pipeline/manual/replay-hourly-new-users.ts
#!/usr/bin/env npx tsx

import { execFileSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { executeD1, queryD1 } from "../shared/d1.ts";
import { buildEmailFilter, loadEmailCohort } from "../shared/email-cohort.ts";

type Environment = "staging";

interface ParsedArgs {
    env: Environment;
    emailsFile: string;
    skipPrepare: boolean;
    hourlyDryRun: boolean;
}

const scriptDir = dirname(fileURLToPath(import.meta.url));
const workspaceRoot = resolve(scriptDir, "../../..");
const repoRoot = resolve(workspaceRoot, "..");
const dotenvPath = resolve(repoRoot, ".env");

function parseArguments(): ParsedArgs {
    const args = process.argv.slice(2);
    const getString = (flag: string, fallback = ""): string => {
        const index = args.indexOf(flag);
        return index >= 0 && args[index + 1] ? args[index + 1] : fallback;
    };

    const env = getString("--env", "staging");
    if (env !== "staging") {
        console.error(`❌ Unsupported --env ${env}`);
        process.exit(1);
    }

    const emailsFile = getString("--emails-file");
    if (!emailsFile) {
        console.error("❌ --emails-file is required");
        process.exit(1);
    }

    return {
        env: "staging",
        emailsFile,
        skipPrepare: args.includes("--skip-prepare"),
        hourlyDryRun: args.includes("--hourly-dry-run"),
    };
}

function countCohortUsers(env: Environment, emails: string[]): number {
    const rows = queryD1(
        env,
        `SELECT COUNT(*) AS count FROM user WHERE 1=1${buildEmailFilter("email", emails)}`,
    );
    return Number(rows[0]?.count ?? 0);
}

function prepareCohort(env: Environment, emails: string[]): void {
    const ok = executeD1(
        env,
        `UPDATE user SET tier = 'microbe', tier_balance = 0, trust_score = NULL, score = NULL, score_checked_at = NULL, banned = 0, ban_reason = NULL, ban_expires = NULL WHERE 1=1${buildEmailFilter("email", emails)}`,
    );
    if (!ok) {
        throw new Error("Failed to prepare hourly replay cohort in D1");
    }
}

function printSummary(env: Environment, emails: string[]): void {
    const rows = queryD1(
        env,
        `SELECT
            SUM(CASE WHEN tier = 'microbe' THEN 1 ELSE 0 END) AS microbe_count,
            SUM(CASE WHEN tier = 'spore' THEN 1 ELSE 0 END) AS spore_count,
            SUM(CASE WHEN tier = 'seed' THEN 1 ELSE 0 END) AS seed_count,
            SUM(CASE WHEN COALESCE(banned, 0) = 1 THEN 1 ELSE 0 END) AS banned_count,
            SUM(CASE WHEN trust_score >= 60 THEN 1 ELSE 0 END) AS trusted_count,
            SUM(CASE WHEN trust_score < 60 AND trust_score IS NOT NULL THEN 1 ELSE 0 END) AS blocked_count
         FROM user
         WHERE 1=1${buildEmailFilter("email", emails)}`,
    );
    const row = rows[0] ?? {};
    console.log("\n📊 Cohort summary:");
    console.log(`   Microbe: ${row.microbe_count ?? 0}`);
    console.log(`   Spore: ${row.spore_count ?? 0}`);
    console.log(`   Seed: ${row.seed_count ?? 0}`);
    console.log(`   Banned: ${row.banned_count ?? 0}`);
    console.log(`   Trust >= 60: ${row.trusted_count ?? 0}`);
    console.log(`   Trust < 60: ${row.blocked_count ?? 0}`);
}

function loadDotenvEnv(): NodeJS.ProcessEnv {
    const env = { ...process.env };
    if (!existsSync(dotenvPath)) {
        return env;
    }

    for (const rawLine of readFileSync(dotenvPath, "utf-8").split(/\r?\n/)) {
        const line = rawLine.trim();
        if (!line || line.startsWith("#") || !line.includes("=")) continue;
        const [key, ...rest] = line.split("=");
        const rawValue = rest.join("=").trim();
        const value =
            rawValue.startsWith('"') && rawValue.endsWith('"')
                ? rawValue.slice(1, -1)
                : rawValue.startsWith("'") && rawValue.endsWith("'")
                  ? rawValue.slice(1, -1)
                  : rawValue;
        if (!(key in env)) {
            env[key] = value;
        }
    }

    return env;
}

function runCommand(args: string[], env: NodeJS.ProcessEnv): void {
    execFileSync("npm", args, {
        cwd: workspaceRoot,
        env,
        stdio: "inherit",
    });
}

function main(): void {
    const config = parseArguments();
    const emails = loadEmailCohort(config.emailsFile);
    if (!emails) {
        throw new Error("No emails loaded for replay");
    }

    const cohortSize = countCohortUsers(config.env, emails);
    console.log("🧪 Replay Hourly New-User Pipeline");
    console.log(`   Environment: ${config.env}`);
    console.log(`   Cohort file: ${config.emailsFile}`);
    console.log(`   Cohort size: ${cohortSize}`);

    if (cohortSize === 0) {
        console.error("❌ No users matched the supplied emails on staging");
        process.exit(1);
    }

    const childEnv = loadDotenvEnv();

    if (!config.skipPrepare) {
        console.log("\n🛠️ Preparing cohort for hourly replay...");
        prepareCohort(config.env, emails);
    } else {
        console.log("\n⏭️ Skipping cohort preparation");
    }

    console.log("\n🔍 Running trust gate...");
    runCommand(
        [
            "run",
            "user-pipeline:trust-score",
            "--",
            "--env",
            config.env,
            "--parallel",
            "3",
            "--store-status",
            "--emails-file",
            config.emailsFile,
        ],
        childEnv,
    );

    console.log("\n🌱 Running hourly new-user pipeline...");
    const command = [
        "run",
        "user-pipeline:hourly-new-users",
        "--",
        "--env",
        config.env,
        "--emails-file",
        config.emailsFile,
    ];
    if (config.hourlyDryRun) {
        command.push("--dry-run");
    }
    runCommand(command, childEnv);

    printSummary(config.env, emails);
}

main();

File: enter.pollinations.ai/scripts/user-pipeline/manual/replay-hourly-new-users.py
#!/usr/bin/env python3
"""Prepare a staging cohort and replay the hourly new-user pipeline on it."""

import argparse
import os
import subprocess
import sys
from pathlib import Path

SCRIPT_ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(SCRIPT_ROOT / "shared"))
from python_runtime import ensure_python_bin

ensure_python_bin()

from d1 import ensure_safe_env, run_d1_query

WORKDIR = Path(__file__).resolve().parents[3]
REPO_ROOT = WORKDIR.parent
DOTENV_PATH = REPO_ROOT / ".env"


def load_email_cohort(file_path: str) -> list[str]:
    try:
        content = Path(file_path).read_text(encoding="utf-8")
    except OSError as error:
        raise RuntimeError(f"Failed to read --emails-file {file_path}: {error}") from error

    emails = list(
        dict.fromkeys(
            line.strip()
            for line in content.splitlines()
            if line.strip() and not line.strip().startswith("#")
        )
    )
    if not emails:
        raise RuntimeError(f"--emails-file {file_path} did not contain any emails.")
    return emails


def build_email_filter(emails: list[str]) -> str:
    values = ", ".join("'" + email.replace("'", "''") + "'" for email in emails)
    return f"email IN ({values})"


def count_cohort_users(env: str, emails: list[str]) -> int:
    results = run_d1_query(
        f"SELECT COUNT(*) AS count FROM user WHERE {build_email_filter(emails)}",
        env,
    )
    if results is None:
        raise RuntimeError("Failed to count replay cohort users in D1")
    if not results:
        return 0
    return int(results[0]["count"])


def prepare_cohort(env: str, emails: list[str]) -> None:
    result = run_d1_query(
        f"""
        UPDATE user
        SET
            tier = 'microbe',
            tier_balance = 0,
            trust_score = NULL,
            score = NULL,
            score_checked_at = NULL,
            banned = 0,
            ban_reason = NULL,
            ban_expires = NULL
        WHERE {build_email_filter(emails)}
        """,
        env,
    )
    if result is None:
        raise RuntimeError("Failed to prepare hourly replay cohort in D1")


def print_summary(env: str, emails: list[str]) -> None:
    results = run_d1_query(
        f"""
        SELECT
            SUM(CASE WHEN tier = 'microbe' THEN 1 ELSE 0 END) AS microbe_count,
            SUM(CASE WHEN tier = 'spore' THEN 1 ELSE 0 END) AS spore_count,
            SUM(CASE WHEN tier = 'seed' THEN 1 ELSE 0 END) AS seed_count,
            SUM(CASE WHEN COALESCE(banned, 0) = 1 THEN 1 ELSE 0 END) AS banned_count,
            SUM(CASE WHEN trust_score >= 60 THEN 1 ELSE 0 END) AS trusted_count,
            SUM(CASE WHEN trust_score < 60 AND trust_score IS NOT NULL THEN 1 ELSE 0 END) AS blocked_count
        FROM user
        WHERE {build_email_filter(emails)}
        """,
        env,
    )
    if not results:
        print("⚠️ No summary rows returned")
        return

    row = results[0]
    print("\n📊 Cohort summary:")
    print(f"   Microbe: {row.get('microbe_count', 0)}")
    print(f"   Spore: {row.get('spore_count', 0)}")
    print(f"   Seed: {row.get('seed_count', 0)}")
    print(f"   Banned: {row.get('banned_count', 0)}")
    print(f"   Trust >= 60: {row.get('trusted_count', 0)}")
    print(f"   Trust < 60: {row.get('blocked_count', 0)}")


def load_dotenv_env() -> dict[str, str]:
    env = os.environ.copy()
    if not DOTENV_PATH.exists():
        return env

    for raw_line in DOTENV_PATH.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        key = key.strip()
        value = value.strip()
        if value.startswith(("'", '"')) and value.endswith(("'", '"')) and len(value) >= 2:
            value = value[1:-1]
        env.setdefault(key, value)

    key_path = env.get("GITHUB_APP_PRIVATE_KEY_PATH")
    if key_path and not Path(key_path).exists():
        repo_pem = next(REPO_ROOT.glob("*.pem"), None)
        if repo_pem is not None:
            env["GITHUB_APP_PRIVATE_KEY_PATH"] = str(repo_pem)
        else:
            env.pop("GITHUB_APP_PRIVATE_KEY_PATH", None)
            env.pop("GITHUB_APP_ID", None)
    return env


def run_command(command: list[str], env: dict[str, str]) -> None:
    subprocess.run(command, cwd=WORKDIR, env=env, check=True)


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Prepare a staging cohort and replay the hourly new-user pipeline"
    )
    parser.add_argument(
        "--env",
        choices=["staging"],
        default="staging",
        help="Environment",
    )
    parser.add_argument(
        "--emails-file",
        required=True,
        help="Newline-separated list of cohort emails",
    )
    parser.add_argument(
        "--skip-prepare",
        action="store_true",
        help="Skip resetting the cohort before replaying the pipeline",
    )
    parser.add_argument(
        "--hourly-dry-run",
        action="store_true",
        help="Run the hourly tier step in dry-run mode after the live trust gate",
    )
    args = parser.parse_args()

    env = ensure_safe_env(args.env)
    emails = load_email_cohort(args.emails_file)
    cohort_size = count_cohort_users(env, emails)

    print("🧪 Replay Hourly New-User Pipeline")
    print(f"   Environment: {env}")
    print(f"   Cohort file: {args.emails_file}")
    print(f"   Cohort size: {cohort_size}")

    if cohort_size == 0:
        print("❌ No users matched the supplied emails on staging", file=sys.stderr)
        return 1

    child_env = load_dotenv_env()

    if not args.skip_prepare:
        print("\n🛠️ Preparing cohort for hourly replay...")
        prepare_cohort(env, emails)
    else:
        print("\n⏭️ Skipping cohort preparation")

    print("\n🔍 Running trust gate...")
    run_command(
        [
            "npm",
            "run",
            "user-pipeline:trust-score",
            "--",
            "--env",
            env,
            "--parallel",
            "3",
            "--store-status",
            "--emails-file",
            args.emails_file,
        ],
        child_env,
    )

    print("\n🌱 Running hourly new-user pipeline...")
    command = [
        "npm",
        "run",
        "user-pipeline:hourly-new-users",
        "--",
        "--env",
        env,
        "--emails-file",
        args.emails_file,
    ]
    if args.hourly_dry_run:
        command.append("--dry-run")
    run_command(command, child_env)

    print_summary(env, emails)
    return 0


if __name__ == "__main__":
    sys.exit(main())
